File: 2019/csharp/Day1.cs
namespace AdventOfCode2019;

class Day1 : RobotElf
{
    public Day1() : base(1) {}

    public override long Part1()
    {
        return Input.Select(long.Parse).Select(x => Math.Max((x / 3) - 2, 0)).Sum();
    }

    public override long Part2() 
    {
        Func<int, int> f = x => Math.Max((x / 3) - 2, 0);
        Func<int, int> g = x => {
            var s = 0;
            while (x > 0) 
            {
                x = f(x);
                s += x;
            }
            return s;
        };

        return Input.Select(int.Parse).Select(g).Sum();
    }
}

File: 2019/fsharp/day1.fs
namespace AdventOfCode.FSharp.Y2019

open System

// The Tyranny of the Rocket Equation
module Day1 =

    let part1 (input : string seq) =
        let f x = max ((x / 3) - 2) 0
        let q = input |> Seq.map Int32.Parse |> Seq.map f
        q |> Seq.sum

    let part2 (input : string seq) = 
        let f x  = max ((x / 3) - 2) 0
        let g x =
            let y = f x 
            if y > 0 then Some(y, y) else None
        let h x = (List.unfold g x) |> List.sum

        input |> Seq.map Int32.Parse |> Seq.map h |> Seq.sum
